tests/difference/core/quint_model/ccv.qnt

```diff
@@ -0,0 +1,825 @@
+        consumersToStart: Set[Chain],
+        consumersToStop: Set[Chain],
+        consumersToTimeout: Set[Chain]
+    ): (Chain -> str, str) = {
+        // check if any consumer is both started and stopped
+        if (consumersToStart.intersect(consumersToStop).size() > 0) {
+            (currentConsumerStatusMap, "Cannot start and stop a consumer at the same time")
+        } else {
+            val res1 = currentConsumerStatusMap.startConsumers(consumersToStart)
+            val newConsumerStatus = res1._1
+            val err1 = res1._2
+            val res2 = newConsumerStatus.stopConsumers(consumersToStop, consumersToTimeout)
+            val err2 = res2._2
+            if (err1 != "") {
+                (currentConsumerStatusMap, err1)
+            } else if (err2 != "") {
+                (currentConsumerStatusMap, err2)
+            } else {
+                (res2._1, "")
+            }
+        }
+    }
+
+
+    // Takes the currentValidatorSet and puts it as the newest set of the voting history
+    pure def enterCurValSetIntoBlock(chainState: ChainState): ChainState = {
+        chainState.with(
+            "votingPowerHistory", chainState.votingPowerHistory.prepend(
+                chainState.currentValidatorSet
+            )
+        )
+    }
+
+    // Advances the timestamp in the chainState by timeAdvancement
+    pure def advanceTime(chainState: ChainState, timeAdvancement: Time): ChainState = {
+        chainState.with(
+            "lastTimestamp", chainState.lastTimestamp + timeAdvancement
+        )
+    }
+
+    // common logic to update the chain state, used by both provider and consumers.
+    pure def endAndBeginBlockShared(chainState: ChainState, timeAdvancement: Time): ChainState = {
+        chainState.enterCurValSetIntoBlock().advanceTime(timeAdvancement)
+    }
+
+    // returns the providerState with the following modifications:
+    // * sends VscPackets to all running consumers
+    // * increments the runningVscId
+    // This should only be called when the provider chain is ending a block,
+    // and only when the running validator set is considered to have changed
+    // and there is a consumer to send a packet to.
+    pure def sendVscPackets(providerState: ProviderState): ProviderState = {
+        val newSentPacketsPerConsumer = ConsumerChains.mapBy(
+            (consumer) => 
+            // if validator set changed and the consumer is running, send a packet
+            if (providerState.providerValidatorSetChangedInThisBlock and 
+                isRunningConsumer(consumer, providerState)) {
+                List({
+                        id: providerState.runningVscId,
+                        validatorSet: providerState.chainState.currentValidatorSet,
+                          sendingTime: providerState.chainState.lastTimestamp
+                })
+            } else {
+                List()
+            }
+        )
+        providerState.with(
```

Comment:
would not it be nicer to introduce values for the new syntax and then use the `{ ... }` syntax, e.g., `{ ...providerState, outstandingPacketsToConsumer: newOPC, sentVscPacketsToConsumer: newSPC }` ?

Reply:
I changed it here and in a few other places where the .with semantics was used a bit too much. I agree the ... way is usually nicer, I just didn't think of it while writing this :)